Add Search Panel for Genes

mikaell opened the pull request:
This PR adds a functionality: Search genes in panels

Reference: https://github.com/Clinical-Genomics/scout/issues/3497


* Search is verbatim
* No regex or case conversion is done
* Searches genes.hgnc_id and genes.symbol in Panel collection and returns all results


<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the paxa software installed on `hasta`:
 - Log in into hasta: `ssh <USER.NAME>@hasta.scilifelab.se`
 - Activate the staging environment with the command `us` (Use Stage)
 - Run the command `paxa` and follow the instructions. Make sure to specify scout-stage as the resource you request and the server cg-vm1 as server.
1. Log out from the hasta server.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, log out from `cg-vm1` and log in again in the `hasta` server, repeat the `hasta` and `paxa` procedure, which will release the allocated resource (scout-stage) to be used for testing by other users.
</details>


**How to test**:
1. how to test it, possibly with real cases/data2. 

Goto `http://localhost:5000/panels`.

Search for a gene symbol or hgnc id. If present in panel database it should be returned together with panel version.


![image](https://user-images.githubusercontent.com/1150065/179693621-76ccacfe-1184-4a74-b122-b2e0114d5c4b.png)

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

mikaell commented:
Leaving it here awaiting feedback.

northwestwitch commented on CHANGELOG.md:
```diff
@@ -17,6 +17,7 @@ About changelog [here](https://keepachangelog.com/en/1.0.0/)
 - Better pagination and number of variants returned by queries in `Search SNVs and INDELs` page
 - Refactored and simplified code used for collecting gene variants for `Search SNVs and INDELs` page
 - Fix sidebar panel icons in Case view
+- Ensure genome builds are input to Scout as strings from config files and CLI input
```
Mmm looks like this line belongs to another PR?

mikaell replied:
Fixed

northwestwitch commented on scout/build/case.py:
```diff
@@ -1,7 +1,7 @@
 from datetime import datetime
 
-from scout.constants import CUSTOM_CASE_REPORTS
+from scout.constants import CUSTOM_CASE_REPORTS, GENOME_BUILD_37, GENOME_BUILD_38
```
This is also part or another branch perhaps?

mikaell replied:
Fixed

northwestwitch commented on scout/commands/load/case.py:
```diff
@@ -109,11 +110,11 @@ def case(
 
-    if config_data.get("genome_build") not in [37, 38, "37", "38"]:
-        config_data["genome_build"] = int(
+    if config_data.get("genome_build") not in [GENOME_BUILD_37, GENOME_BUILD_38]:
```
Same here

mikaell replied:
Hmmm… Must have been some mix-up. Fixing now.

northwestwitch commented on scout/server/blueprints/panels/views.py:
```diff
@@ -37,7 +47,22 @@ def api_panels(panel_name):
 def panels():
     """Show all panels for a user"""
-    if request.method == "POST":  # Edit/create a new panel and redirect to its page
+    LOG.debug("Show all panels for a user")
```
```suggestion
```

northwestwitch commented on scout/server/blueprints/panels/views.py:
```diff
@@ -37,7 +47,22 @@ def api_panels(panel_name):
     """Show all panels for a user"""
-    if request.method == "POST":  # Edit/create a new panel and redirect to its page
+    LOG.debug("Show all panels for a user")
+    LOG.debug("REQUEST: {} {} {}".format(request.args, request.data, request.form))
```
```suggestion
```

northwestwitch commented on scout/server/blueprints/panels/views.py:
```diff
@@ -37,7 +47,22 @@ def api_panels(panel_name):
+    if request.method == "POST" and not request.form.get(
+        "search_for"
+    ):  # Edit/create a new panel and redirect to its page
+        LOG.debug("request.method == 'POST'")
```
```suggestion
```

northwestwitch commented on tests/commands/load/test_load_case_cmd.py:
```diff
@@ -3,6 +3,7 @@
 import tempfile
 
 from scout.commands import cli
+from scout.constants import GENOME_BUILD_37
```
Ops, here again!

mikaell replied:
Fixed

mikaell replied:
`% grep -rn GENOME_BUILD_37 *` returns nothing now.

northwestwitch commented on scout/server/blueprints/panels/templates/panels/panels.html:
```diff
@@ -27,6 +28,48 @@
+<div class="container-float">
+  <div class="card mt-3">
+    <div class="card-body">
+      <form action="{{ url_for('panels.panels') }}" enctype="multipart/form-data" method="POST">
```
enctype="multipart/form-data"  is generally used if the form contains a file upload. Since this is not the case I think it would be best to omit the enctype so it will be the default (application/x-www-form-urlencoded).

mikaell replied:
Removed!

northwestwitch commented on scout/server/blueprints/panels/templates/panels/panels.html:
```diff
@@ -27,6 +28,48 @@
+      <form action="{{ url_for('panels.panels') }}" enctype="multipart/form-data" method="POST">
+        <div class="row">
+          <div class="col-5">
+            <input type="text" name="search_for" class="form-control" placeholder="Search Gene in Panels" required
```
Have you checked how the gene search is performed on other pages, for instance on the genes page?

The search there is more user friendly than a verbatim search because when you enter 3 letters or more it suggests you the possible genes, by matching the entered letters in gene names and also aliases:

<img width="859" alt="image" src="https://user-images.githubusercontent.com/28093618/183578770-327bc0f1-aeca-491f-9e6c-d8227de2aa65.png">

What happens in your code:

<img width="745" alt="image" src="https://user-images.githubusercontent.com/28093618/183578836-00692dde-3527-4e39-b414-66169354cd5e.png">

But you don't know if the gene is not found because you spelled it wrong or because it changed official name and it became an alias or because it doesn't exist in Scout.

mikaell replied:
Typeahead is copied from genes-page.

northwestwitch commented on scout/server/blueprints/panels/templates/panels/panels.html:
```diff
@@ -27,6 +28,48 @@
+          <div class="col">
+            <button type="submit" class="btn btn-secondary">Search!</button>
+          </div>
+          <div class="col">
```
Remove this empty col perhaps?

mikaell replied:
Done!

northwestwitch commented on scout/parse/models.py:
```diff
@@ -282,7 +282,7 @@ class ScoutLoadConfig(BaseModel):
     gene_fusion_report_research: Optional[str] = None
     gene_panels: Optional[List[str]] = []
-    genome_build: int = Field(38, alias="human_genome_build")
+    genome_build: str = Field(38, alias="human_genome_build")
```
Here also, this comes from another branch

mikaell replied:
Fixed.

northwestwitch commented on scout/server/blueprints/panels/templates/panels/panels.html:
```diff
@@ -27,6 +28,48 @@
+      <div id="SearchResults" class="card-body">
+        {% if search_result %}
+        <strong> <em>{{search_string}}</em> found in: </strong> <br>
+        {% for name, version in search_result %}
```
Search results are quite difficult to read, especially if a gene is found on many panels and panel versions:

<img width="709" alt="image" src="https://user-images.githubusercontent.com/28093618/183579781-53e92f35-7a12-49ab-bd8a-e9ebc11e7a6a.png">

Would be nice to apply some formatting to have these results in a table or something

mikaell replied:
Done!

northwestwitch commented on scout/server/blueprints/panels/templates/panels/panels.html:
```diff
@@ -66,7 +109,7 @@
         <div class="row">
           <div class="col text-center">
-            <button type="submit" class="btn btn-secondary">Upload</button>
+            <button type="submit" class="btn btn-secondary">UploadS</button>
```
```suggestion
            <button type="submit" class="btn btn-secondary">Upload</button>
```

northwestwitch commented on scout/server/blueprints/panels/views.py:
```diff
@@ -73,6 +98,8 @@ def panels():
         panel_names=panel_names,
         panel_versions=panel_versions,
         institutes=institutes,
+        search_string=search_string,
```
This string doesn't show in the search form after performing the search. Perhaps avoid passing it to the template?

northwestwitch replied:
Ahh, I was looking into the form field, to be precompiled with the term from the previous search!

northwestwitch commented on scout/server/blueprints/panels/views.py:
```diff
@@ -37,7 +47,22 @@ def api_panels(panel_name):
+        panels_found = store.search_panels(search_string)
+
+    # Add new panel
+    if request.method == "POST" and not request.form.get(
```
```suggestion
    elif request.method == "POST":
```

northwestwitch reviewed (changes requested):
It appears something went wrong with pulls and pushes and this PR is a mix of two distinct branches (search gene in panels and genome builds branch). 

Nice search panel in the gene panels page! The search is working of one enters the exact symbol or a gene, but I think the feature would benefit a a lot from a more user friendly search (with search suggestions), as I wrote in one of my comment.

Also the presentation of the search result might be improved, given that very often multiple panels and panel version present the same gene and as the code is now it shows a long list of  results that are not super easy to read. For instance consider a table or a <ul> element with distinct colors for each distinct gene panel. Or something! 😄

mikaell commented:
TODOs
- [ ] Make search more user friendly (with search suggestions)
- [ ] Format search results for readability

northwestwitch commented:
Let me know when you are done with the fixes on this one @mikaell, so I'll give it another round!

mikaell commented:
> Let me know when you are done with the fixes on this one @mikaell, so I'll give it another round!

I think it is ready. I've implemented a better display of search results and suggestions in the search field. Was there anything else?

northwestwitch commented:
> I think it is ready. I've implemented a better display of search results and suggestions in the search field. Was there anything else?

There's a couple of comments of mine that you haven't seen (easy things)

mikaell commented:
> > I think it is ready. I've implemented a better display of search results and suggestions in the search field. Was there anything else?
> 
> There's a couple of comments of mine that you haven't seen (easy things)

Now I think I've fixed all the comments.

northwestwitch commented:
> Now I think I've fixed all the comments.

Thanks, I'll look at it!

northwestwitch commented on scout/adapter/mongo/panel.py:
```diff
@@ -549,23 +549,36 @@ def latest_panels(self, institute_id, include_hidden=False):
+            search_int = search_string
+        query = {"$or": [{"genes.hgnc_id": search_int}, {"genes.symbol": search_string}]}
+        result = self.panel_collection.find(query)
+        result_list = [[element["panel_name"], element["version"]] for element in result]
```
What you'll get after this query is this: ['name', 1.0, 'name', 2.0]

But what you need to display results is this: [['name', [1.0, 2.0]]]

And to reformat results you have created 2 new functions (_compile_panel_versions and _compile_panel_versions) below. 

You don't need all that code if you change your query to return just what you need instead. Try using an aggregation:

```
[{
    $match: {
        'genes.hgnc_id': <HGNC-iD>
    }
}, {
    $group: {
        _id: {
            display_name: '$display_name'
        },
        versions: {
            $addToSet: '$version'
        }
    }
}, {
    $sort: {
        display_name: 1
    }
}]
```

That will return results like this:

<img width="727" alt="image" src="https://user-images.githubusercontent.com/28093618/185884425-19bf1941-eda7-46f2-b5ef-0509b9ab293a.png">

northwestwitch commented on scout/server/blueprints/cases/views.py:
```diff
@@ -439,7 +387,14 @@ def phenotypes_actions(institute_id, case_name):
+                "Provided gene info could not be parsed! " "Please allow autocompletion to finish.",
+                "warning",
+            )
+        # TODO: dont crash
```
```suggestion
```

mikaell replied:
That's always a good reminder, but I'll remove it.

northwestwitch commented on scout/server/blueprints/panels/templates/panels/panels.html:
```diff
@@ -27,6 +28,68 @@
+
+
+ 
+<!-- Create a Bootstrap card containing text bar and submit button for searching -->
```
This looks really nice now!

northwestwitch commented on scout/server/blueprints/cases/views.py:
```diff
@@ -439,7 +387,14 @@ def phenotypes_actions(institute_id, case_name):
+            hgnc_ids = parse_raw_gene_ids(request.form.getlist("genes"))
+        except ValueError:
+            flash(
+                "Provided gene info could not be parsed! " "Please allow autocompletion to finish.",
```
In which case does this crash? If I don't type something that it likes I just get this validation error:

<img width="552" alt="image" src="https://user-images.githubusercontent.com/28093618/185885582-3f077bb5-28b2-4dc2-a2ea-bb4fe0b2c10c.png">

mikaell replied:
It's possible, I've managed. But don't remember how. 

Anyways, the function `parse_raw_gene_ids()` was extracted from cases/views. Here it had the side-effect of flashing a warning. When moving the functionality to utils/genes.py the side-effect of flashing messages had to be extracted. Therefor to keep the same behaviour the try/except now is outside and kept the same. cases.views.py and panels.views.py will be similar.

northwestwitch commented on scout/server/blueprints/panels/templates/panels/panels.html:
```diff
@@ -219,5 +287,26 @@ <h5 class="modal-title">Remove gene panel "{{panel.display_name}}"?</h5>
+        minLength: 3,
+      });
+
+      $('select[multiple]').selectpicker({
```
I don't think you need this one: you don't have any multiple selects in the form of this page

mikaell replied:
I'm not an expert in javascript. But I interpret `$('select[multiple]').selectpicker'` as being a dynamic index, not allowing for multiple selects in a dropdown. Changing the code to `$('select[i]').selectpicker` has no apparent effect on the page's behaviour.

mikaell replied:
Well, removing it completely changed nothing either. 

Done.

northwestwitch commented on scout/server/blueprints/panels/views.py:
```diff
@@ -37,7 +48,34 @@ def api_panels(panel_name):
+                "warning",
+            )
+        # TODO: dont crash after error
+        LOG.debug("SEARCH: {}".format(search_string))
```
```suggestion
```

northwestwitch commented on scout/server/blueprints/panels/views.py:
```diff
@@ -37,7 +48,34 @@ def api_panels(panel_name):
+            )
+        # TODO: dont crash after error
+        LOG.debug("SEARCH: {}".format(search_string))
+        LOG.debug("HGNC: {}".format(hgnc_symbols))
```
```suggestion
```

northwestwitch commented on scout/server/blueprints/panels/views.py:
```diff
@@ -37,7 +48,34 @@ def api_panels(panel_name):
+
+        for hgnc_id in hgnc_symbols:
+            panels_found = store.search_panels(hgnc_id)
+            panels_found_compiled.extend(_compile_panel_versions(panels_found))
```
You don't need this if you optimize the query!

mikaell replied:
I don't understand which line you are referring to?

northwestwitch replied:
I mean that is you optimize the query you don't need _compile_panel_versions and _compile_panel_versions_aux, the adapter's search_panels function can return the list of panel in the right format

mikaell replied:
Ok.

But I assume that the a for-look is still necessary as `parse_raw_gene_symbols(raw_symbols_list)` may return a set with several entries?


```
def parse_raw_gene_symbols(raw_symbols_list):
    """Parse list of concatenated gene symbol list for hgnc_symbols from Phenomizer.

    Arguments:
        raw_symbols(list(string)) - e.g. ("POT1 | MUTYH", "POT1 | ATXN1 | ATXN7")

    Returns:
        hgnc_symbols(set(string)) - set of (unique) gene symbols without intervening chars
    """

```

northwestwitch replied:
Yes, I guess so!

northwestwitch commented on scout/server/blueprints/panels/views.py:
```diff
@@ -311,3 +351,42 @@ def gene_edit(panel_id, hgnc_id):
     return dict(panel=panel_obj, form=form, gene=hgnc_gene, panel_gene=panel_gene)
+
+
+def _compile_panel_versions(panels_found):
```
You don't need these 2 functions. Just optimize the query instead

mikaell replied:
Done!

northwestwitch commented on scout/server/blueprints/panels/views.py:
```diff
@@ -37,7 +48,34 @@ def api_panels(panel_name):
+        LOG.debug("SEARCH: {}".format(search_string))
+        LOG.debug("HGNC: {}".format(hgnc_symbols))
+
+        for hgnc_id in hgnc_symbols:
```
This is not actual any more since you are using the autocompletion no?

If you try searching by more that one symbol the form doesn't submit:

<img width="572" alt="image" src="https://user-images.githubusercontent.com/28093618/185887082-915c7d30-c893-4dc0-b735-3588c8643a47.png">

<img width="522" alt="image" src="https://user-images.githubusercontent.com/28093618/185887163-607919d5-f3c5-4b4f-a089-8428ff774a03.png">

mikaell replied:
Right now search is only supported for one gene at a time. Though the input check is done in javascript, the backend is not prepared to handle multiple search words separated by commas. 

-Do we/should we support this? 


![image](https://user-images.githubusercontent.com/1150065/185890678-e46c71e4-d6ae-44ce-84ae-160f94575e19.png)

northwestwitch commented on scout/server/blueprints/panels/views.py:
```diff
@@ -37,7 +48,34 @@ def api_panels(panel_name):
+            LOG.debug("FOUND: {}".format(panels_found_compiled))
+    # Add new panel
+    elif request.method == "POST":
+        "search_for"
```
What's this, a comment?

mikaell replied:
A random string 😃

northwestwitch reviewed (changes requested):
The looks of the results are much better now and the autocompletion works, but I think it still needs some polishing.

I have a few comments, the most important thing being optimizing the query to avoid a lot of processing of the query results downstream.

mikaell commented on scout/adapter/mongo/panel.py:
```diff
@@ -549,23 +549,36 @@ def latest_panels(self, institute_id, include_hidden=False):
+        return set(item["_id"] for item in query_result)
+
+    def search_panels(self, search_string):
+        """Return all panels and versions that contain given gene, list is sorted"""
```
Done!

northwestwitch reviewed (changes requested):
Much improved! I still have a few minor comments.

Take also a look at what is flagged by SonarCloud, perhaps some of these things are easily fixable!

mikaell commented on scout/adapter/mongo/panel.py:
```diff
@@ -549,23 +549,59 @@ def latest_panels(self, institute_id, include_hidden=False):
-        return set(item["_id"] for item in query)
+        return set(item["_id"] for item in query_result)
+
+    def search_panels(self, search_string):
```
It's a pretty good API function.

northwestwitch replied:
But it's not used, and returns a result very similar to the other, but formatted worse

mikaell replied:
OK

mikaell commented on scout/adapter/mongo/panel.py:
```diff
@@ -549,23 +549,59 @@ def latest_panels(self, institute_id, include_hidden=False):
+        query = [
+            {"$match": {"genes.hgnc_id": hgnc_id}},
+            {"$group": {"_id": "$display_name", "versions": {"$addToSet": "$version"}}},
+            {"$sort": {"display_name": 1}},
```
Fixed.

mikaell commented on scout/server/blueprints/cases/views.py:
```diff
@@ -439,7 +387,13 @@ def phenotypes_actions(institute_id, case_name):
+            hgnc_ids = parse_raw_gene_ids(request.form.getlist("genes"))
+        except ValueError:
+            flash(
+                "Provided gene info could not be parsed! " "Please allow autocompletion to finish.",
```
I don't trust users 🙂

I've made this crash, but can't reproduce it a work. Maybe I could do it on my home machine because it is so slow and unresponsive on some webpages.

northwestwitch commented on scout/adapter/mongo/panel.py:
```diff
@@ -549,23 +549,49 @@ def latest_panels(self, institute_id, include_hidden=False):
+                }
+            },
+            {"$unwind": "$versions"},
+            {"$sort": {"_id": 1, "versions": 1}},
```
🥇

northwestwitch reviewed (approved):
Looks good now! 💯 

<img width="720" alt="image" src="https://user-images.githubusercontent.com/28093618/186335088-e61d79b2-6d9b-48c0-80f9-19b29b40909f.png">

Feel free to merge. It's going to be a very useful feature!
